### PR TITLE
Use MPI for management in `cupyx.distributed`

### PR DESCRIPTION
**Pull request by emcastillo**
This is needed by #6228
We will use MPI to efficiently exchange the sizes of sparse matrices because doing it with NCCL causes device synchronization and the TCP store will kill performance when used heavily.

### CODE REVIEW

File: cupyx/distributed/_nccl_comm.py [hunk 1]
```diff
@@ -44,25 +51,49 @@ class NCCLBackend(_Backend):
...
+            # make them positive and send them as bytes to the proxy store
+            shifted_nccl_id = bytes([b + 128 for b in nccl_id])
+            self._store_proxy['nccl_id'] = shifted_nccl_id
             self._store_proxy.barrier()
         else:
             self._store_proxy.barrier()
             nccl_id = self._store_proxy['nccl_id']
-        # Initialize devices
-        nccl_id = tuple([int(b) - 128 for b in nccl_id])
+            nccl_id = tuple([int(b) - 128 for b in nccl_id])
```
Comment (leofang) [inline cupyx/distributed/_nccl_comm.py:96]: `nccl_id` is gone for rank 0?

File: tests/cupyx_tests/distributed_tests/comm_runner.py [hunk 2]
```diff
@@ -46,91 +46,118 @@ def run_broadcast(rank, n_workers, root, dtype):
...
-    _launch_workers(N_WORKERS, run_broadcast, (0, dtype))
-    _launch_workers(N_WORKERS, run_broadcast, (1, dtype))
+    if use_mpi:
+        from mpi4py import MPI
+        # This process was run with mpiexec
+        run_broadcast(MPI.COMM_WORLD.Get_rank(), 0, dtype, False)
+        run_broadcast(MPI.COMM_WORLD.Get_rank(), 1, dtype, False)
+    else:
+        _launch_workers(run_broadcast, (0, dtype))
+        _launch_workers(run_broadcast, (1, dtype))
```
Comment (leofang) [inline tests/cupyx_tests/distributed_tests/comm_runner.py:56]: Can't we do this inside `_launch_workers` to avoid code dup/branching? Something like
```
def _launch_workers(func, args, ..., use_mpi=False):
    if use_mpi:
        from mpi4py import MPI
        func(...)
    else:
        # the old _launch_workers impl
```

Comment (leofang) [inline tests/cupyx_tests/distributed_tests/comm_runner.py:56]: Can't we do this inside `_launch_workers` to avoid code dup/branching? Something like
```python
def _launch_workers(func, args, ..., use_mpi=False):
    if use_mpi:
        from mpi4py import MPI
        func(...)
    else:
        # the old _launch_workers impl
```

Comment (emcastillo) [reply tests/cupyx_tests/distributed_tests/comm_runner.py:56]: I tried, but the code was hard to follow and the only repeated lines are 2, 
```
nccl_id = nccl.get_unique_id()
...
self._comm = nccl.NcclCommunicator(n_devices, nccl_id, rank)
```
and for the older case since we need to manipulate the nccl_id it becomes quite of spaghetti code with ifs ...
I think this approach is much simpler to read

**Review by kmaehashi — APPROVED** [review]
